common/txmgr/inmemory_store.go

```diff
@@ -0,0 +1,1869 @@
+}
+
+// UpdateTxAttemptInProgressToBroadcast updates a transaction attempt from in_progress to broadcast.
+// It also updates the transaction state to unconfirmed.
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) UpdateTxAttemptInProgressToBroadcast(
+	ctx context.Context,
+	tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+	attempt txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+	newAttemptState txmgrtypes.TxAttemptState,
+) error {
+	if tx.BroadcastAt == nil {
+		return fmt.Errorf("update_tx_attempt_in_progress_to_broadcast: unconfirmed transaction must have broadcast_at time")
+	}
+	if tx.InitialBroadcastAt == nil {
+		return fmt.Errorf("update_tx_attempt_in_progress_to_broadcast: unconfirmed transaction must have initial_broadcast_at time")
+	}
+	if tx.State != TxInProgress {
+		return fmt.Errorf("update_tx_attempt_in_progress_to_broadcast: can only transition to unconfirmed from in_progress, transaction is currently %s", tx.State)
+	}
+	if attempt.State != txmgrtypes.TxAttemptInProgress {
+		return fmt.Errorf("update_tx_attempt_in_progress_to_broadcast: attempt must be in in_progress state")
+	}
+	if newAttemptState != txmgrtypes.TxAttemptBroadcast {
+		return fmt.Errorf("update_tx_attempt_in_progress_to_broadcast: new attempt state must be broadcast, got: %s", newAttemptState)
+	}
+
+	ms.addressStatesLock.RLock()
+	defer ms.addressStatesLock.RUnlock()
+	as, ok := ms.addressStates[tx.FromAddress]
+	if !ok {
+		return fmt.Errorf("update_tx_attempt_in_progress_to_broadcast: %w", ErrAddressNotFound)
+	}
+
+	// Persist to persistent storage
+	if err := ms.txStore.UpdateTxAttemptInProgressToBroadcast(ctx, tx, attempt, newAttemptState); err != nil {
+		return fmt.Errorf("update_tx_attempt_in_progress_to_broadcast: %w", err)
+	}
+
+	// Update in memory store
+	if err := as.MoveInProgressToUnconfirmed(*tx, attempt); err != nil {
+		return fmt.Errorf("update_tx_attempt_in_progress_to_broadcast: %w", err)
+	}
+
+	return nil
+}
+
+// FindNextUnstartedTransactionFromAddress returns the next unstarted transaction for a given address.
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) FindNextUnstartedTransactionFromAddress(_ context.Context, tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], fromAddress ADDR, chainID CHAIN_ID) error {
+	if ms.chainID.String() != chainID.String() {
+		return fmt.Errorf("find_next_unstarted_transaction_from_address: %w", ErrInvalidChainID)
+	}
+	ms.addressStatesLock.RLock()
+	defer ms.addressStatesLock.RUnlock()
+	as, ok := ms.addressStates[fromAddress]
+	if !ok {
+		return fmt.Errorf("find_next_unstarted_transaction_from_address: %w", ErrAddressNotFound)
+	}
+
+	// ensure that the address is not already busy with a transaction in progress
+	if as.inprogress != nil {
+		return fmt.Errorf("find_next_unstarted_transaction_from_address: address %s is already busy with a transaction in progress", fromAddress)
+	}
```

Comment:
Since this is just a Find function, I don't think we need validations like this in the memory store. I feel like the responsibility to check this should be with the Broadcaster